sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -510,69 +514,38 @@ public void teardown() throws Exception {
     void setParameters(T element, PreparedStatement preparedStatement) throws Exception;
   }
 
-  /** A {@link PTransform} to write to a JDBC datasource. */
-  @AutoValue
-  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+  /** Abstract base class for JdbcIO write operations. */
+  abstract static class AbstractWrite<RowT, InputT> extends PTransform<PCollection<InputT>, PDone> {
     @Nullable abstract DataSourceConfiguration getDataSourceConfiguration();
     @Nullable abstract String getStatement();
+    @Nullable abstract PreparedStatementSetter<RowT> getPreparedStatementSetter();
     abstract long getBatchSize();
-    @Nullable abstract PreparedStatementSetter<T> getPreparedStatementSetter();
-
-    abstract Builder<T> toBuilder();
-
-    @AutoValue.Builder
-    abstract static class Builder<T> {
-      abstract Builder<T> setDataSourceConfiguration(DataSourceConfiguration config);
-      abstract Builder<T> setStatement(String statement);
-      abstract Builder<T> setBatchSize(long batchSize);
-      abstract Builder<T> setPreparedStatementSetter(PreparedStatementSetter<T> setter);
-
-      abstract Write<T> build();
-    }
-
-    public Write<T> withDataSourceConfiguration(DataSourceConfiguration config) {
-      return toBuilder().setDataSourceConfiguration(config).build();
-    }
-    public Write<T> withStatement(String statement) {
-      return toBuilder().setStatement(statement).build();
-    }
-    public Write<T> withPreparedStatementSetter(PreparedStatementSetter<T> setter) {
-      return toBuilder().setPreparedStatementSetter(setter).build();
-    }
-
-    /**
-     * Provide a maximum size in number of SQL statenebt for the batch. Default is 1000.
-     *
-     * @param batchSize maximum batch size in number of statements
-     * @return the {@link Write} with connection batch size set
-     */
-    public Write withBatchSize(long batchSize) {
-      checkArgument(batchSize > 0, "batchSize must be > 0, but was %d", batchSize);
-      return toBuilder().setBatchSize(batchSize).build();
-    }
 
     @Override
-    public PDone expand(PCollection<T> input) {
+    public PDone expand(PCollection<InputT> input) {
       checkArgument(
           getDataSourceConfiguration() != null, "withDataSourceConfiguration() is required");
       checkArgument(getStatement() != null, "withStatement() is required");
       checkArgument(
           getPreparedStatementSetter() != null, "withPreparedStatementSetter() is required");
 
-      input.apply(ParDo.of(new WriteFn<T>(this)));
+      input.apply(ParDo.of(createWriteFn(this)));
       return PDone.in(input.getPipeline());
     }
 
-    private static class WriteFn<T> extends DoFn<T, Void> {
+    protected abstract AbstractWriteFn<RowT, InputT>
+            createWriteFn(AbstractWrite<RowT, InputT> write);
 
-      private final Write<T> spec;
+    abstract static class AbstractWriteFn<RowT, InputT> extends DoFn<InputT, Void> {
```

Comment:
Can this be achieved with composition rather than inheritance? (same about the enclosing transform) - composition is generally much easier to deal with.